File: go/internal/noise/frame_test.go
// go/internal/noise/frame_test.go
package noise

import (
	"bytes"
	"testing"
)

func TestDataFrameRoundTrip(t *testing.T) {
	enc := EncodeData([]byte("ls -la\n"))
	typ, payload, err := DecodeFrame(enc)
	if err != nil {
		t.Fatal(err)
	}
	if typ != FrameData {
		t.Fatalf("expected FrameData, got %d", typ)
	}
	if !bytes.Equal(payload, []byte("ls -la\n")) {
		t.Fatalf("payload mismatch: %q", payload)
	}
}

func TestResizeFrameRoundTrip(t *testing.T) {
	enc := EncodeResize(120, 40)
	typ, payload, err := DecodeFrame(enc)
	if err != nil {
		t.Fatal(err)
	}
	if typ != FrameResize {
		t.Fatalf("expected FrameResize, got %d", typ)
	}
	cols, rows, err := DecodeResize(payload)
	if err != nil {
		t.Fatal(err)
	}
	if cols != 120 || rows != 40 {
		t.Fatalf("expected 120x40, got %dx%d", cols, rows)
	}
}

func TestWindowsFrameRoundTrip(t *testing.T) {
	j := []byte(`{"v":1,"active":"@0","win":[{"id":"@0","i":0,"n":"main"}]}`)
	enc := EncodeWindows(j)
	if enc[0] != byte(FrameWindows) {
		t.Fatalf("expected FrameWindows tag 0x04, got %#x", enc[0])
	}
	typ, payload, err := DecodeFrame(enc)
	if err != nil {
		t.Fatal(err)
	}
	if typ != FrameWindows || !bytes.Equal(payload, j) {
		t.Fatalf("windows frame mismatch: type=%d payload=%q", typ, payload)
	}
}

func TestControlFrameRoundTrip(t *testing.T) {
	j := []byte(`{"a":"select-window","t":"@3"}`)
	enc := EncodeControl(j)
	if enc[0] != byte(FrameControl) || byte(FrameControl) != 0x05 {
		t.Fatalf("expected FrameControl tag 0x05, got %#x", enc[0])
	}
	typ, payload, err := DecodeFrame(enc)
	if err != nil || typ != FrameControl || !bytes.Equal(payload, j) {
		t.Fatalf("control frame mismatch")
	}
}

func TestDecodeEmptyFrameErrors(t *testing.T) {
	if _, _, err := DecodeFrame(nil); err == nil {
		t.Fatal("expected error decoding empty frame")
	}
}

File: web/test/frame.test.js
// web/test/frame.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
  FRAME_DATA,
  FRAME_RESIZE,
  FRAME_WINDOWS,
  FRAME_CONTROL,
  encodeData,
  encodeResize,
  encodeWindows,
  encodeControl,
  decodeFrame,
  decodeResize,
} from '../src/noise/frame.js';

test('data frame round trip', () => {
  const enc = encodeData(new TextEncoder().encode('ls -la\n'));
  const { type, payload } = decodeFrame(enc);
  assert.equal(type, FRAME_DATA);
  assert.equal(new TextDecoder().decode(payload), 'ls -la\n');
});

test('resize frame round trip', () => {
  const enc = encodeResize(120, 40);
  const { type, payload } = decodeFrame(enc);
  assert.equal(type, FRAME_RESIZE);
  const { cols, rows } = decodeResize(payload);
  assert.equal(cols, 120);
  assert.equal(rows, 40);
});

test('windows frame round trip (byte-mirrors Go)', () => {
  const j = new TextEncoder().encode('{"v":1,"active":"@0","win":[{"id":"@0","i":0,"n":"main"}]}');
  const enc = encodeWindows(j);
  assert.equal(enc[0], FRAME_WINDOWS);
  assert.equal(enc[0], 0x04);
  const { type, payload } = decodeFrame(enc);
  assert.equal(type, FRAME_WINDOWS);
  assert.deepEqual(payload, j);
});

test('control frame round trip (byte-mirrors Go)', () => {
  const j = new TextEncoder().encode('{"a":"select-window","t":"@3"}');
  const enc = encodeControl(j);
  assert.equal(enc[0], FRAME_CONTROL);
  assert.equal(enc[0], 0x05);
  const { type, payload } = decodeFrame(enc);
  assert.equal(type, FRAME_CONTROL);
  assert.deepEqual(payload, j);
});

test('decoding empty frame throws', () => {
  assert.throws(() => decodeFrame(new Uint8Array(0)));
});
